ntbs-integration-tests/Helpers/WebApplicationExtensions.cs

```diff
@@ -0,0 +1,81 @@
+using System.Collections.Generic;
+using System.Net.Http;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc.Testing;
+using Microsoft.AspNetCore.TestHost;
+using Microsoft.Extensions.DependencyInjection;
+using ntbs_integration_tests.TestServices;
+using ntbs_service;
+using ntbs_service.Models;
+using ntbs_service.Services;
+
+namespace ntbs_integration_tests.Helpers
+{
+    public static class WebApplicationExtensions
+    {
+        public static WebApplicationFactory<Startup> WithNhsUserBuilder(this WebApplicationFactory<Startup> factory, IDictionary<int, string> idServiceCodeMapping)
+        {
+            return factory.WithWebHostBuilder(builder =>
+            {
+                SetupDatabase(builder, idServiceCodeMapping);
+                builder.ConfigureTestServices(services =>
+                {
+                    services.AddScoped<IUserService, NhsUserService>();
+                });
+            });
+        }
+
+        public static WebApplicationFactory<Startup> WithPheUserBuilder(this WebApplicationFactory<Startup> factory,
+                                                                        IDictionary<int, string> idServiceCodeMapping,
+                                                                        IDictionary<int, string> idPostcodeMapping)

```

Comment:
These methods are a bit of a tangle at first glance.
The decision whether to have `NHS` user or `PHE` user seems to be unnecessarily coupled with inserting reference data into the db here. What do you think about having the syntax be more along the lines of:

```
 factory
  .WithNhsUser()
  .WithNotificationAndTbServicesConnected(idServiceCodeMapping) // or .WithNotificationAndTbServiceConnected(notificationId, tbServiceCode)
  .WithoutRedirects() // or .CreateClient(), as per other comment in this file
```

Reply:
@jasiekmiko I think under the hood that would mean chaining multiple "WithWebHostBuilder" methods together (as both the db setup and the service setup are happening within the existing "WithWebHostBuilder" extension method) - is that ok? Not sure whether they can be chained like that

Reply:
Ok chainign WithWebHostBuilder seems to be fine :)
I agree that we're coupling things in a weird way here. I'd go with splitting it into just two extension methods, one generic one for setting the mock user service, and one for setting the data (as we might want to set both postcodes and tb services at the same time for certain tests, so I think one data setting method makes sense)

Reply:
Ah, awesome!

Splitting into just two is definitely an improvement, but I'm a little worried about this pattern growing: if there's a need to add more and more bits that should be seeded this way. I imagined a scenario where you can chain multiple of these db setups if needed, too, like:
```
 factory
  .WithNhsUser()
  .WithNotificationAndTbServiceConnected(notificationId, tbServiceCode)
  .WithNotificationAndTbServiceConnected(notificationId2, tbServiceCode2)
  .CreateClient()
```
though i appreciate that might a bit of repetition in getting the scope and db connections..

Reply:
yeah, that's what I was trying to avoid. However, ultimately it will only slow down the tests a bit but make it a lot more readable - so I think we should just go with your suggestion!